server/src/api/v1.py

```diff
@@ -652,3 +656,45 @@ def queue_wait_time_percentiles_get():
             queue["wait_times"]
         )
     return queue_percentile_data
+
+
+def generate_token(permissions, secret_key):
+    """Generates JWT token with queue permission given a secret key"""
+    expiration_time = datetime.utcnow() + timedelta(seconds=2)
+    token_payload = {
+        "exp": expiration_time,
+        "iat": datetime.now(timezone.utc),  # Issued at time
+        "sub": "access_token",
+        "permissions": permissions,
+    }
+    token = jwt.encode(token_payload, secret_key, algorithm="HS256")
+    return token
+
+
+def validate_client_key_pair(client_id: str, client_key: str):
+    """
+    Checks client_id and key pair for validity and returns their permissions
+    """
+    client_permissions_entry = database.mongo.db.client_permissions.find_one(
+        {
+            "client_id": client_id,
+            "client_secret_hash": client_key,
+        }
+    )
+    if client_permissions_entry is None:
+        return {}
+
+    permissions = client_permissions_entry["permissions"]
+    return permissions
+
+
+SECRET_KEY = secrets.token_hex(20)
+
+
+@v1.get("/authenticate/token/<client_id>")
```

Comment:
If there's a 1:1 correspondence between the `client_id` and the `client_key` then you don't really need both; the key in the header should be enough to retrieve the permissions entry.

Reply:
In this case, I think having both the id and a key/token/secret (the private part of the credentials) is useful so that you don't have to search through all the known IDs to see if it matches one of them. I would imagine there could be another use here where it's not 1:1 at all, but rather a user being able to generate multiple API keys, giving them names, expirations, etc. In that case, you would definitely want some sort of ID to compare against.